### ISSUE TITLE
Add jsonpath assertion

### ISSUE THREAD

**Issue by gwcary** [sassoftware/unravl#20]
I'd like to see a more straightforward assert/extractor using json path expressions:

    "assert": [
        {"status": 200},
        {"jsonpath" : "$.links[0].rel == userPreferences"},
        {"jsonpath" : "$.links[0].method == GET"},
        {"jsonpath" : "$.links[0].type == application/vnd.sas.collection"},
        {"jsonpath" : "$.links[0].uri == /preferences/preferences/stpweb1"},
        {"jsonpath" : "$.links[1].rel == createPreferences"},
        {"jsonpath" : "$.links[1].method == POST"},
        {"jsonpath" : "$.links[1].type == application/vnd.sas.collection"},
        {"jsonpath" : "$.links[1].uri == /preferences/preferences/stpweb1"}
    ]

This uses the library  'com.jayway.jsonpath:json-path:2.+'

**Comment by DavidBiesack** [comment]
Thanks for the suggestion. I have a couple suggestions and comments.

See the existing proposal for a [jsonPath](https://github.com/sassoftware/unravl/blob/master/doc/Assertions-to-do.md#jsonpath) assertion which is similar to what you suggest but would (if implemented) be closer in style to existing UnRAVL assertions. It would work with numeric, string, boolean, or JSON object/array values. Your suggestion is a little ambiguous because I think it implies the right hand side is always an unquoted string value. I would like to avoid complicated new expression grammars such as strings that contain "*jsonPath == value*". UnRAVL tries to use JSON structures (objects and arrays) to avoid having to parse values. 

Thus, my preference if implementing a jsonPath assertion would be to use the existing [jsonPath](https://github.com/sassoftware/unravl/blob/master/doc/Assertions-to-do.md#jsonpath) proposal.

Another option I've considered is a json structural match such as [JSONassert](https://github.com/skyscreamer/JSONassert) which would let you compare a JSON result to an expected JSON structure. However, JSONassert does not do complete subset comparison (only on objects). For your example, a more UnRAVL-like assertion would be:
```
"assert" : { "json" :
              { "links" : 
                [{
                  "rel" : "userPreferences",
                  "method" : "GET",
                  "type" : "application/vnd.sas.collection",
                  "uri" : "/preferences/preferences/stpweb1"
                 },
                 {
                  "rel" : "createPreferences",
                  "method" : "POST",
                  "type" : "application/vnd.sas.collection",
                  "uri" : "/preferences/preferences/stpweb1"
                 }
                 ]
             },
             "contains" : true,
             "ignoreArrayOrder" : true
          }
```
which would compare the JSON response to the in-line JSON, but only do a containment comparison, not an exact comparison - that is, the response contains a "links"
object that is an array, and that array contains the two link objects listed (in any order).

Also, look at [Links](https://github.com/sassoftware/unravl/blob/master/doc/Bind.md#links-and-hrefs) which allows simple extraction of links from a JSON response. You could then assert that extracted link objects match JSON 
```
"bind" : { "links" : [ "userPreferences", "createPreferences" ] },
"assert" : [ "userPreferences == expectedUserPreferences",
                 "createPreferences == exectedCreatePreferences"
               ]
```
Using the links extractor makes the test more resilient to server-side changes in the order of links. Your example will break if a new link is added between link 0 and link 1.

**Comment by DavidBiesack** [comment]
See #21

**Comment by gwcary** [comment]
The problem with your proposal is that it doesn't allow for a very rich set
of comparators.  It seems to assume you're always going to want
equivalence, or at least containment of a fairly coarse json document.
There are going to be cases where I'd like to compare things on a more
fine-grained level, and how would you compare against an integer using a
greater than, less than,etc?  There needs to be greater flexibility for
including relational operators.  You can add more syntactic elements to
include relational operands, but I think you'll end up with a more
complicated grammar, rather than less complicated.  I think, to do what I
want requires three elements, the JSON path, the element to compare, and a
relational operator.  Maybe it needs to be something like this:

{jsonPath": [
     {"path" : "$.links[0].rel", "comparison" : "==", "value" :
"userPreferences"},
     {"path" : "$.links[*].method", "comparison" : "==", "value" : ["GET",
"POST"]},
     {"path" : "$.version", "comparison" : ">=", "value" : 1}
]

This is a little more unwieldy than I'd like, but it does allow a richer
set of comparisons, and allows more than just strings for comparison values.

**Comment by DavidBiesack** [comment]
Another option is to define jsonPath as an extractor that simply binds variables, perhaps slightly more elegantly than the "groovy" extractor which requires use of textValue() etc. Then use "groovy" assertions to do arbitrary comparisons if you want more than equality or containment. I think most validations will be based on equality so the jsonPath assertion would cover most cases.

**Comment by DavidBiesack** [comment]
Merged pull request #25 to implement jsonPath `"bind"` element. You can then use Groovy assertions on extracted values.
